solver/include/solver/generic_butcher_rk.hpp

```diff
@@ -57,6 +60,204 @@ namespace ode::butcher
             double tol;
         };
 
+        namespace detail
+        {
+            template <typename T>
+            concept has_identity_method = std::is_member_function_pointer_v<decltype( &T::identity )>;
+
+            template <typename T>
+            concept has_solver_method = std::is_member_function_pointer_v<decltype( &T::solver )>;
+
+            template <typename T, typename... Args>
+            concept has_newton_method = std::is_member_function_pointer_v<decltype( &T::template newton<Args...> )>;
+        } // namespace detail
+
+        template <typename state_t>
+            requires std::ranges::range<state_t>
+        auto
+        norm( state_t const& x )
+        {
+            auto r = static_cast<decltype( *std::begin( x ) )>( 0. );
+            for ( auto const& xi : x )
+            {
+                r += std::abs( xi ) * std::abs( xi );
+            }
+
+            return std::sqrt( r );
+        }
+
+        template <typename state_t>
+        auto
+        norm( state_t const& x )
+        {
+            return std::abs( x );
+        }
+
+        template <typename value_t, typename state_t, typename func_t, typename jacobian_t, typename solver_t>
+        state_t
+        newton( func_t&& f, jacobian_t&& df, state_t const& x0, solver_t&& solver, value_t tol = 1e-10, std::size_t max_iter = 50 )
+        {
+            state_t xk             = x0;
+            value_t res            = norm( f( xk ) );
+            std::size_t count_iter = 0;
+
+            while ( count_iter < max_iter && res > tol )
+            {
+                auto increment = solver( df( xk ), -f( xk ) );
+
+                xk  = xk + increment;
+                res = norm( f( xk ) );
+
+                count_iter += 1;
+            }
+
+            return xk;
+        }
+
+        template <typename Tableau, typename LinearAlgebra_t = void>
+        struct diagonal_implicit_rk_butcher
+        {
+            Tableau butcher;
+            static constexpr std::size_t N_stages     = Tableau::N_stages;
+            static constexpr bool is_embedded         = is_embedded_tableau<Tableau>;
+            static constexpr std::size_t order        = Tableau::order;
+            static constexpr char* id                 = Tableau::id;
+            static constexpr bool void_linear_algebra = std::is_void<LinearAlgebra_t>::value;
+            using linear_algebra_t                    = typename std::conditional<void_linear_algebra,
```

Comment:
Why do you  have two definitions of `linear_alegebra_t`? see [here](https://github.com/hpc-maths/ponio/pull/22/files#diff-138980c591060deff2877eadb7558bf65a9ef1807237a4b11052ba961974580fR160)